### PR TITLE
Testing

### PR DESCRIPTION
**Pull request by ufocoder**

### CODE REVIEW

File: services/fhir.ts [hunk 1]
```diff
@@ -216,7 +216,7 @@ export function makeReference<T extends AidboxResource>(
 
 export function isReference<T extends AidboxResource>(
     resource: T | AidboxReference<T>
-): resource is AidboxReference<T> {
+): boolean {
```
Comment (ruscoder) [inline services/fhir.ts]: Why did you change the type guard here?

Comment (ufocoder) [reply services/fhir.ts]: `_.isEmpty` returns boolean
https://lodash.com/docs/4.17.15#isEmpty

File: __tests__/hooks/pager.spec.ts [hunk 2]
```diff
@@ -0,0 +1,139 @@
...
+    const resourcesOnPage = 2;
+
+    beforeEach(() => {
+        jest.clearAllMocks();
+    });
+
+    describe('property `hasNext`', () => {
+        test("returns false when there's no next page", () => {
+            const data = success({
+                id: 'fakeID',
```
Comment (ruscoder) [inline __tests__/hooks/pager.spec.ts]: Let's make tests in a more FHIR way. It will help us easily read the tests.
e.g. data = success({ resourceType: 'Bundle' }) without id (bundle is a virtual resource and used for list returning as a result

File: __tests__/hooks/pager.spec.ts [hunk 3]
```diff
@@ -0,0 +1,139 @@
...
+
+            expect(hasNext).toBeFalsy();
+            expect(remoteData).toEqual(data);
+        });
+
+        test("returns true when there's next page", () => {
+            const data = success({
+                id: 'fakeID',
+                resourceType: 'type',
+                link: [
```
Comment (ruscoder) [inline __tests__/hooks/pager.spec.ts]: Here is a real example of Bundle.link
"link":[{"relation":"first","url":"/EpisodeOfCare?page=1"},{"relation":"self","url":"/EpisodeOfCare?page=1"}]

File: __tests__/hooks/pager.spec.ts [hunk 4]
```diff
@@ -0,0 +1,139 @@
...
+    test('method `loadNext`', async () => {
+        const data = success({
+            id: 'fakeID',
+            resourceType: 'type',
+        });
+
+        (<jest.Mock>useService).mockImplementation(() => [data]);
+
+        const { result } = renderHook(() => usePager<User>(resourceType, resourcesOnPage));
+        const { loadNext } = result.current[1];
```
Comment (ruscoder) [inline __tests__/hooks/pager.spec.ts]: Too hard to read.
Let's rewrite:
```
const [remoteData, { loadNext} ] = result.current;

```

File: __tests__/hooks/pager.spec.ts [hunk 5]
```diff
@@ -0,0 +1,139 @@
...
+            expect(pageToLoad).toBe(1);
+            expect(fhirSearchParams).toEqual({ _count: 2, _page: 1 });
+            expect(fhirResourceType).toEqual(resourceType);
+        }
+
+        act(() => {
+            loadNext();
+        });
+
+        {
```
Comment (ruscoder) [inline __tests__/hooks/pager.spec.ts]: Why it is wrapped into curly braces?

File: __tests__/services/fhir.spec.ts [hunk 6]
```diff
@@ -0,0 +1,305 @@
...
+
+describe('Service `fhir`', () => {
+    beforeEach(() => {
+        jest.clearAllMocks();
+    });
+
+    test('method `getFHIRResource`', async () => {
+        const reference = {
+            id: '1',
+            resourceType: 'type',
```
Comment (ruscoder) [inline __tests__/services/fhir.spec.ts]: use 'Patient' as universal resourceType. This resource has no mandatory fields and very good for testing purposes.

File: services/fhir.ts (hunk 1 above)
Comment (ir4y) [reply services/fhir.ts]: Indeed @ufocoder, but here we need type predicate behavior.
https://www.typescriptlang.org/docs/handbook/advanced-types.html#type-guards-and-differentiating-types

The same technique is using for RemoteData type https://github.com/beda-software/aidbox-react/blob/master/libs/remoteData.ts#L51

File: __tests__/hooks/crud.spec.ts [hunk 7]
```diff
@@ -0,0 +1,293 @@
...
+                    status: 'Success',
+                    data: defaultData,
+                });
+            });
+        });
+
+        describe('has correct behavior with parameter `defaultResource`', () => {
+            const id = 'toggle';
+            const getOrCreate = true;
+            const defaultResource = {
```
Comment (ir4y) [inline __tests__/hooks/crud.spec.ts]: const defaultResource:Patient = {
                resourceType,
                active: false,
            };

File: __tests__/hooks/crud.spec.ts [hunk 8]
```diff
@@ -0,0 +1,293 @@
...
+                    data: 'data',
+                });
+            });
+
+            test('when FHIR response has error', async () => {
+                (<jest.Mock>getFHIRResource).mockImplementation(() => failure('error'));
+
+                const { result, waitForNextUpdate } = renderHook(() => useCRUD<Patient>(resourceType, id, getOrCreate));
+
+                expect(result.current[0]).toEqual({
```
Comment (ir4y) [inline __tests__/hooks/crud.spec.ts]: Use data constructor instead

```
expect(result.current[0]).toEqual(loading)
```

File: __tests__/hooks/crud.spec.ts [hunk 9]
```diff
@@ -0,0 +1,293 @@
...
+            handleDelete(resource);
+        });
+
+        expect(result.current[0]).toEqual({
+            status: 'Loading',
+        });
+
+        await waitForNextUpdate();
+
+        expect(result.current[0]).toEqual({
```
Comment (ir4y) [inline __tests__/hooks/crud.spec.ts]: Let's also check that related service `deleteFHIRResource` was called.

Comment (ir4y) [reply __tests__/hooks/crud.spec.ts]: This also makes sense for any other CRUD operations from `CRUDManager`.

File: __tests__/hooks/pager.spec.ts (hunk 4 above)
Comment (ir4y) [reply __tests__/hooks/pager.spec.ts]: At https://github.com/beda-software/aidbox-react/blob/3805ee3c10fccdbfd8f3411629cb66dc92724a42/__tests__/hooks/crud.spec.ts#L33-L37
you use destructuring



It makes sense to extract values in the same way in all cases.

File: __tests__/hooks/service.spec.ts [hunk 10]
```diff
@@ -16,7 +16,8 @@ describe('Hook `useService`', () => {
     };
 
     test('change `Loading` to `Success` status when service resolved', async () => {
-        const { result, waitForNextUpdate } = renderHook(() => useService(service));
+        const deps = [1, 2];
```
Comment (ir4y) [inline __tests__/hooks/service.spec.ts]: We also need a test that checks that changing deps cause service rerun.
